### PR TITLE
Fixes challenges titles

### PR DESCRIPTION
**Pull request by lili2311**
### ISSUE/TICKET NUMBER
[ISSUE_130](https://trello.com/c/MTEg1H1j/130-challenge-hierarchy-page-without-challenges-added-design-implementation)
[ISSUE_130](https://trello.com/c/MTEg1H1j/130-challenge-hierarchy-page-without-challenges-added-design-implementation)

### Describe the changes
- Added titles to all pages
- Formatted the challenges view and renamd some vars
- Amended the challenges template to render a design that matches the wireframes
- put the button styles into their own file

### Screenshots (if appropriate)
screen shot 2017-11-27 at 23 15 01
screen shot 2017-11-27 at 23 15 08
screen shot 2017-11-27 at 23 17 15

### Questions or comments (if any)
Regarding the front end tody up ticket, please let me know what is unclear/confusing and I can document it, this project already has very bare CSS.

### CODE REVIEW

**Comment by gnocchi2815** [comment]
The changes made To the view have changed the sorting of the challenges. There is only ever one challenge in progress, the sorting by initial anxiety level should be for the challenges not in progress but this PR has removed that and now uses lambda to sort the challenge in progress which isn’t necessary

**Comment by gnocchi2815** [comment]
Just to add to my comment above the variable in the view should be named challenge_in_progress (not challenges_in_progress) for better readability as there can only be one challenge in progress.

**Comment by gnocchi2815** [comment]
The view is now getting the in progress challenge twice

File: ocdaction/challenges/views.py [hunk 1]
```diff
@@ -27,23 +27,19 @@ def challenge_list(request):
...
     )[:10]
 
-    for challenge in challenges_not_in_progress:
-        print(challenge.in_progress)
-
     challenges_in_progress_sorted = sorted(
         challenges_in_progress.all(),
         reverse=True,
         key=lambda c: int(c.get_latest_initial_anxiety_level())
     )
```
Comment (gnocchi2815) [inline ocdaction/challenges/views.py]: This should be sorting the challenges not in progress:
`   challenges_not_in_progress_sorted = sorted(
        challenges_not_in_progress.all(),
        reverse=True,
        key=lambda c: int(c.get_latest_initial_anxiety_level())
    )`

File: ocdaction/challenges/views.py [hunk 2]
```diff
@@ -27,23 +27,19 @@ def challenge_list(request):
...
 
-    # challenge_in_progress = Challenge.objects.filter(in_progress=True)
-    anxiety_score_card = AnxietyScoreCard.objects.filter(challenge=challenges_in_progress).last()
+    challenge_in_progress = Challenge.objects.filter(in_progress=True)
+    anxiety_score_card = AnxietyScoreCard.objects.filter(challenge=challenge_in_progress).last()
```
Comment (gnocchi2815) [inline ocdaction/challenges/views.py]: Already have the challenge in progress at the start of the review so line 36 not required

File: ocdaction/challenges/views.py [hunk 3]
```diff
@@ -27,23 +27,19 @@ def challenge_list(request):
...
 
     context = {
         'challenges_in_progess': challenges_in_progress_sorted,
         'challenges_not_in_progress': challenges_not_in_progress,
-        # 'another_challenge_in_progress': challenge_in_progress,
```
Comment (gnocchi2815) [inline ocdaction/challenges/views.py]: Context changes
`        'challenge_in_progess': challenge_in_progress,`
and
`        'challenges_not_in_progress': challenges_not_in_progress_sorted,`

obviously then will need to make change to variable name in template

**Comment by mybits** [comment]
@Lili 
Following @gnocchi2815 comments: as a rule of thumb let's not introduce backend changes in solely frontend tickets unless:
- it's refactoring (hence no changes in behaviour) and the changes are committed in a separate commits clearly explaining why they are done

When changes impact behaviours:
- let's first discuss it with the team to avoid any re-work and after agreement let's do them in a separate ticket to enhance transparency and clarity for everyone (also for any future developers :))

In general let's try to apply this rule to any changes that are extras to the ticket/user story - let's always have them in a separate commit with the explanation of 'why' and 'what' 👌

**Comment by gnocchi2815** [comment]
I will look in more detail later when on laptop but I think there is also an issue here with the view and challenge list page. The original view and code prevented a user from starting a challenge if there is already one in progress.

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 4]
```diff
@@ -1,48 +1,94 @@
...
-                    </ul>
-                {% else %}
-                    <ul class="col-xs-6 col-xs-offset-3">
-                        <li class="row">
-                            Level: {% get_anxiety_level challenge %}
-                            <a href="{% url 'challenge' challenge.id %}"><h3>{{ challenge.challenge_name }}</h3></a>
-                            <a class="btn btn-primary pull-right col-xs-3" href="{% url 'challenge-score-form-new' challenge.id %}"> Do the challenge </a>
-                        </li>
-                    </ul>
-                {% endif %}
```
Comment (gnocchi2815) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: Lines 24-40 in the original code checked whether another challenge was in progress and if so, disabled the "do the challenge" button for the rest of the challenges

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 5]
```diff
@@ -51,28 +45,27 @@ <h3 class="text-align-center">In progress</h3>
...
                                         <tr>
-                                            <td class="text-align-left">{% get_anxiety_level challenge %}</td>
-                                            <td class="text-align-left"><a href="{% url 'challenge' challenge.id %}">{{ challenge.challenge_name }}</a></td>
-                                            <td><a class="" href="">Do challenge</a></td>
+                                            <td class="text-align-left ibox-content">{% get_anxiety_level challenge %}</td>
+                                            <td class="text-align-left ibox-content"><a href="{% url 'challenge' challenge.id %}">{{ challenge.challenge_name | title }}</a></td>
+                                            <td class="ibox-content text-align-right"><a class="" href="{% url 'challenge-score-form-new' challenge.id %}"><span class="glyphicon glyphicon-play-circle margin-right-5"></span>Do the challenge</a></td>
                                         </tr>
                                     {% endfor %}
                                 </tbody>
```
Comment (gnocchi2815) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: I have added a comment to the previous commit to highlight the code that was removed but which is required. For the challenges not in progress it should check whether another challenge is in progress.  If there is another challenge in progress then the user cannot start another challenge and the do the challenge button is disabled for all the other challenges.
See PR #112

File: frontend/src/sass/partials/_helpers.scss [hunk 6]
```diff
@@ -14,6 +14,42 @@
...
+}
+
+.unstyled-list {
+    margin: 0 !important;
+    padding: 0 !important;
+
+    li {
+        list-style: none !important;
+    }
+}
```
Comment (mybits) [inline frontend/src/sass/partials/_helpers.scss]: What is the reason of so extensive use of `!important` rule here?

File: ocdaction/ocdaction/templates/challenge/challenge_add.html [hunk 7]
```diff
@@ -1,17 +1,19 @@
 {% extends "layout.html" %}
 
+{% block title %}Act: Add new challenge{% endblock %} 
+
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_add.html]: We don't use 'Act' in challenges context anymore

File: ocdaction/ocdaction/templates/challenge/challenge_add.html [hunk 8]
```diff
@@ -1,17 +1,19 @@
...
 {% load static widget_tweaks %}
 
 {% block main %}
 	<section class="container">
 	    <div class="row">
             <div class="col-xs-offset-3">
                 <a href="{% url 'challenge-list' %}">
-                    <span class="glyphicon glyphicon-menu-left">Back</span>
+                    <span class="glyphicon glyphicon-menu-left">Back to challenges</span>
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_add.html]: In design it's 'Back' (no extras needed)

File: ocdaction/ocdaction/templates/challenge/challenge_add.html [hunk 9]
```diff
@@ -1,17 +1,19 @@
...
                 </a>
             </div>
 			<div class="col-xs-12">
-				<h1> ACT </h1>
+				<h1>ACT</h1>
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_add.html]: This heading is no needed in the new design

File: ocdaction/ocdaction/templates/challenge/challenge_edit.html [hunk 10]
```diff
@@ -1,7 +1,7 @@
 {% extends "layout.html" %}
 {% load static widget_tweaks %}
 
-{% block title %}Edit challenge{% endblock %}
+{% block title %}Act: Edit challenge{% endblock %}
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_edit.html]: Same as above, no 'Act' in use any longer

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 11]
```diff
@@ -1,48 +1,94 @@
 {% extends "layout.html" %}
 
 {% load static %}
+
+{% block title %}Act: My Challenge Hierarchy{% endblock %}
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: And here too

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 12]
```diff
@@ -1,48 +1,94 @@
...
+                                    <th>Challenge</th>
+                                    <th></th>
+                                </tr>
+                                </thead>
+                                <tbody>
+                                    {% for challenge in challenges_not_in_progress %}
+                                        <tr>
+                                            <td class="text-align-left">{% get_anxiety_level challenge %}</td>
+                                            <td class="text-align-left"><a href="{% url 'challenge' challenge.id %}">{{ challenge.challenge_name }}</a></td>
+                                            <td><a class="" href="">Do challenge</a></td>
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: We don't use 'Do the challenge' (or 'task') anymore as it doesn't reflect the exposure cycle. Just an arrow, like in the design, will be enough.

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 13]
```diff
@@ -1,48 +1,94 @@
...
+                                </tbody>
+                            </table>
+                        </div>
+                    </div>
+                </div>
+            {% endif %}
+        {% else %}
+       
+            <div class="custom-panel background--primary text-align-center col-xs-12">
+                <h2 class="h4 margin-top-0">Add a new Challenge to get started</h2>
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: small copy change: 'Add new Challenge to get started'

File: ocdaction/ocdaction/templates/challenge/challenge_list_archived.html [hunk 14]
```diff
@@ -4,17 +4,47 @@
...
+                                {% endfor %}
+                            </tbody>
+                        </table>
+                    </div>
+                </div>
+            </div>
         {% else %}
-            <h2>You have no archived challenges yet</h2>
+    
+            <div class="custom-panel background--primary text-align-center col-xs-12">
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list_archived.html]: I can't actually see it rendered yet, but from the code I reckon it's the green colour? (`background--primary`). In Archive list we use beige-ish one: https://app.zeplin.io/project/59dbd75b74d032e1682a78fc/screen/59dbd84774d032e1682a7ee0

But actually an Archive list it's not the part of this ticket changes so you don't have to implement this :)

File: ocdaction/ocdaction/templates/layout.html [hunk 15]
```diff
@@ -21,12 +21,12 @@
...
-						<li class="header__item">
-							<a class="header__link" href="{% url 'challenge-list-archived' %}">Archive</a>
-						</li>
+    						<li class="header__item">
+    							<a class="header__link" href="{% url 'challenge-list-archived' %}">Archive</a>
+    						</li>
 					        <li class="header__item">
 								<a href="{% url 'logout' %}">
-									<i class="fa fa-sign-out"></i> Log out
+									<i class="fa fa-sign-out"></i>Log out ({{user.username}})
```
Comment (mybits) [inline ocdaction/ocdaction/templates/layout.html]: No need for a username here

**Comment by lili2311** [comment]
Looking at rebasing and fixing the changes now, just to confirm why backend changes were made: 
- made the changes since rendering the template was difficult, it is much cleaner and nicer to separate the in progress form not since they have to be rendered separately. Will mention it in the future upfront thought! Sorry

File: ocdaction/ocdaction/templates/challenge/challenge_add.html (hunk 7 above)
Comment (lili2311) [reply ocdaction/ocdaction/templates/challenge/challenge_add.html]: Should I remove the h1 then that says ACT?

File: frontend/src/sass/partials/_helpers.scss (hunk 6 above)
Comment (lili2311) [reply frontend/src/sass/partials/_helpers.scss]: as these are helpers they need to trump all other styles, so here and if we had base styles would be the only place to do so that is actually appropriate

**Comment by lili2311** [comment]
Updated the PR with requested changes, the CSS is committed so it should actually just show up correctly, in case it still does'nt:
- make sure you have followed the README.md instruction to install npm packages and then grunt + run grunt locally with the default command. Refresh the page and you should see what the screenshots show and what I see locally

**Comment by lili2311** [comment]
- using django title case, just think it looks nicer than letting the user get away with all lower case title names
- will re-check the archive design
- refactoring based on your comments did not work, there would need to be more work done to allow for this as the anonymous lambda functions seems to be giving more data/content to each challenge. So just plucking out the current challenge without using the lamba function is making the django tags error as this challenge doesnt have an associated anxiety score i believe is what it was looking for.

**Comment by gnocchi2815** [comment]
Ah, I see, of course, could we not just capitalise the first word? I think that looks better. Otherwise you can end up with something like "My Car Is Really Dirty And I Don't Want To Touch It" which personally I think isn't quite right - although I agree with its use for titles, buttons etc.
Ok re refactoring, it worked for me locally although perhaps something got lost in translation when I was commenting on specific lines, anyway it all works now so no further issues.

File: frontend/src/sass/components/_navigation.scss [hunk 16]
```diff
@@ -83,7 +83,7 @@
 		margin: 0;
 
 		&:hover {
-			background-color: $ocd-blue-colour;
+			background-color: $ocd-dark-colour
```
Comment (mybits) [inline frontend/src/sass/components/_navigation.scss]: Nitpick: missing semicolon here

File: ocdaction/challenges/views.py [hunk 17]
```diff
@@ -160,12 +167,24 @@ def challenge_score_form_new(request, challenge_id):
             challenge.save()
             anxiety_score_card.save()
 
-            return redirect(challenge_score_form, challenge_id=challenge_id, score_id=anxiety_score_card.id)
+            return redirect(
+                challenge_score_form,
+                challenge_id=challenge_id,
+                score_id=anxiety_score_card.id
+            )
```
Comment (mybits) [inline ocdaction/challenges/views.py]: If it's a formatting changing (which is by the way a very subjective change in this context) can we please do it as a separate commit next time, as it's hard to read if any other changes weren't introduce here

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 18]
```diff
@@ -1,48 +1,107 @@
...
 
 {% block main %}
     <section class="container">
         <h2>My Challenge Hierarchy</h2>
-        <a class="btn btn-primary pull-left col-xs-2 col-xs-offset-5" href="{% url 'challenge-add' %}">
+
+<!--         <a class="btn btn-primary pull-left col-xs-2 col-xs-offset-5" href="{% url 'challenge-add' %}">
             <span class="fa fa-plus-square"></span> Add New Challenge
-        </a>
+        </a> -->
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: not sure why is this commented out?

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 19]
```diff
@@ -1,48 +1,107 @@
...
+                                    <th class="ibox-content bold">Challenge</th>
+                                </tr>
+                                </thead>
+                                <tbody>
+                                    <tr>
+                                        <td class="text-align-left ibox-content">{% get_anxiety_level challenge %}</td>
+                                        <td class="text-align-left ibox-content">
+                                            <a href="{% url 'challenge' challenge.id %}">{{ challenge.challenge_name | title}}</a>
+                                        </td>
+                                        <td class="text-align-right ibox-content"><a class="" href="{% url 'challenge-score-form' challenge.id anxiety_score_card.id %}"><span class="glyphicon glyphicon-hourglass margin-right-5"></span>Continue</a></td>
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: no need for 'Continue' here, just an arrow icon will be enough

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 20]
```diff
@@ -1,48 +1,107 @@
...
+
+                                                {% if another_challenge_in_progress %}
+                                                    <a class="link--disabled">
+                                                        <span class="glyphicon glyphicon-play-circle margin-right-5"></span>
+                                                        Do the challenge
+                                                    </a>
+                                                {% else %}
+                                                    <a href="{% url 'challenge-score-form-new' challenge.id %}">
+                                                        <span class="glyphicon glyphicon-play-circle margin-right-5"></span>
+                                                        Do the challenge
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: Also we don't use 'Do the challenge' wording anymore

**Comment by mybits** [comment]
**Styling comments (the design implementation)**

**Mobile**
screen shot 2017-12-01 at 19 12 59

- can we make the heading to be a normal weight and a bit smaller (to represent the design better) - similar looks to `h2` heading
- a green container shouldn't have any margins but should take the whole screen space
- underline under 'In progress' and 'My Challenges' is not needed
- horizontal lines separating challenges should have 100% width (in mobile)
- no need for a top green border on boxes, but a bottom one (according to design in Zeplin it's grey/silver)
- challenges colour names should be black (same for the icons)
- 'My Challenges' and the 'add' button should be aligned


I think that's mostly it :) As we're taking the mobile first approach no need to focus so much on improvements on a desktop (still many of these changes will apply there as well) 

**General**
- we use Helvetica font across the new design - can that be updated?
- agree with @gnocchi2815 with capitalising all words in challenge fear (name) - no need for that

That's a great progress and great start for the design implementation 👏 👍

File: ocdaction/challenges/views.py (hunk 17 above)
Comment (lili2311) [reply ocdaction/challenges/views.py]: Sure will do, I am using the standard flake8 for python

File: ocdaction/ocdaction/templates/challenge/challenge_list.html (hunk 19 above)
Comment (lili2311) [reply ocdaction/ocdaction/templates/challenge/challenge_list.html]: I will hide the text visually, it is still needed for accessibility. Empty links and buttons are not ok.

File: ocdaction/ocdaction/templates/challenge/challenge_list.html (hunk 20 above)
Comment (lili2311) [reply ocdaction/ocdaction/templates/challenge/challenge_list.html]: What is the wording instead?

**Comment by lili2311** [comment]
I don't think Helvetica is a free font so we cannot use it.

**Comment by lili2311** [comment]
screen shot 2017-12-11 at 23 53 23
screen shot 2017-12-11 at 23 53 38

File: ocdaction/ocdaction/templates/challenge/challenge_list.html (hunk 18 above)
Comment (lili2311) [reply ocdaction/ocdaction/templates/challenge/challenge_list.html]: removed

File: ocdaction/ocdaction/templates/challenge/challenge_list.html [hunk 21]
```diff
@@ -18,26 +18,40 @@ <h2>My Challenge Hierarchy</h2>
...
+                                            </thead>
+                                            <tbody>
+                                                <tr>
+                                                    <td class="text-align-left ibox-content col-xs-3">
+                                                        {% get_anxiety_level challenge %}
+                                                    </td>
+                                                    <td class="text-align-left ibox-content">
+                                                        <a href="{% url 'challenge' challenge.id %}">{{ challenge.challenge_name }}</a>
+                                                    </td>
+                                                    <td class="text-align-right ibox-content"><a href="{% url 'challenge-score-form' challenge.id anxiety_score_card.id %}"><span class="glyphicon glyphicon-hourglass margin-right-5"></span class="visuallyhidden">Continue</a></td>
```
Comment (mybits) [inline ocdaction/ocdaction/templates/challenge/challenge_list.html]: Missing a closing `span` tags. Let's break this line :)
```
<td class="text-align-right ibox-content">
    <a href="{% url 'challenge-score-form' challenge.id anxiety_score_card.id %}">
        <span class="glyphicon glyphicon-play-circle margin-right-5"></span>
        <span class="visuallyhidden">Continue</span>
    </a>
</td>
```

Let's also change the 'hourglass' to a 'play-circle' to follow the design.

**Comment by lili2311** [comment]
Updated

**Comment by mybits** [comment]
LGMT 👍 
@gnocchi2815 could you please review requested changes and if everything is good let's :shipit:
